src/JsonRpcEngine.ts

```diff
@@ -0,0 +1,341 @@
+import SafeEventEmitter from '@metamask/safe-event-emitter';
+import { errorCodes, EthereumRpcError, serializeError } from 'eth-rpc-errors';
+
+/**
+ * A String specifying the version of the JSON-RPC protocol.
+ * MUST be exactly "2.0".
+ */
+export type JsonRpcVersion = '2.0';
+
+/**
+ * An identifier established by the Client that MUST contain a String, Number,
+ * or NULL value if included. If it is not included it is assumed to be a
+ * notification. The value SHOULD normally not be Null and Numbers SHOULD
+ * NOT contain fractional parts.
+ */
+export type JsonRpcId = number | string | void;
+
+export interface JsonRpcError {
+  code: number;
+  message: string;
+  data?: unknown;
+  stack?: unknown;
+}
+
+export interface JsonRpcRequest<T> {
+  jsonrpc: JsonRpcVersion;
+  method: string;
+  id: JsonRpcId;
+  params?: T;
+}
+
+export interface JsonRpcNotification<T> {
+  jsonrpc: JsonRpcVersion;
+  method: string;
+  params?: T;
+}
+
+interface JsonRpcResponseBase {
+  jsonrpc: JsonRpcVersion;
+  id: JsonRpcId;
+}
+
+export interface JsonRpcSuccess<T> extends JsonRpcResponseBase {
+  result: T;
+}
+
+export interface JsonRpcFailure extends JsonRpcResponseBase {
+  error: JsonRpcError;
+}
+
+export type JsonRpcResponse<T> = JsonRpcSuccess<T> | JsonRpcFailure;
+
+export type JsonRpcEngineCallbackError = Error | JsonRpcError | null;
+
+export type JsonRpcEngineReturnHandler = (
+  done: (error?: JsonRpcEngineCallbackError) => void
+) => void;
+
+export type JsonRpcEngineNextCallback = (
+  returnHandlerCallback?: JsonRpcEngineReturnHandler
+) => void;
+
+export type JsonRpcEngineEndCallback = (
+  error?: JsonRpcEngineCallbackError
+) => void;
+
+export type JsonRpcMiddleware = (
+  req: JsonRpcRequest<unknown>,
+  res: JsonRpcResponse<unknown>,
+  next: JsonRpcEngineNextCallback,
+  end: JsonRpcEngineEndCallback
+) => void;
+
+interface InternalJsonRpcResponse extends JsonRpcResponseBase {
```

Comment:
?: could you explain to me the need for distinction between internal and external responses and middleware?

Reply:
If an error is thrown while the response is processed, we preserve the original error as `response._originalError`. That property doesn't exist on the user-facing type, so I had to create an internal type to deal with it. That change then propagated to the middleware type, which also needed an internal equivalent.

I _think_ we can avoid these internal types by passing the error up instead of tacking it on to the `response` object, but I'd rather do that in a follow-up PR.